### PR TITLE
Spike2photom

### PR DESCRIPTION
**Pull request by kouichi-c-nakamura**
This pull request adds an ability to export Photometry continuous data into Spike2 waveform channels.
Only tested for one session so far.
There may be cases where photometry data are missing or specified `photometry_keys` is not found.
Do we want to see some warning or issue an error in such cases?
Take a look at #TODO signs.

Exported Spike2 data itself appeared OK.
There may be an issue if the Photometry file is shorter than the PyControl file. If you can find such a case with an error, I think I can find a way to handle it.

### CODE REVIEW

File: workflows/scripts/03_export_spike2.py [hunk 1]
```diff
@@ -6,21 +6,50 @@
...
 
 #%%
 
 (sinput, soutput) = getSnake(locals(), 'workflows/spout_bar_nov22.smk',
     [settings.debug_folder +'\processed\spike2.smrx'],
   'export_spike2')
 
+#%% Photometry dict
+
+fn = glob(sinput.photometry_folder+'\*.ppd')[0]
```
Comment (teristam) [inline workflows/scripts/03_export_spike2.py:20]: `fn` may be empty if no photometry data is found. We will need to handle the error here and only export the pycontrol data when no photometry data is found

**Review by teristam — CHANGES_REQUESTED** [review]

File: workflows/scripts/03_export_spike2.py (hunk 1 above)
Comment (kouichi-c-nakamura) [reply workflows/scripts/03_export_spike2.py:20]: > Also, suggest to use a Path object to make sure the code can work both in Windows and Linux, e.g. `fn = list(Path(sinput.photometry_folder).glob('*.ppd'))[0]`

Uh-huh, that's the secret of why the random use of file separators is accepted!

**Comment by kouichi-c-nakamura** [comment]
When Photometry signals are saturated, `scale` becomes 0 and when dividing a value with `scale` you get NaN or inf, causing an error when converting float to int16

test run worked

**Comment by kouichi-c-nakamura** [comment]
I made lots of changes because I found that Spike2 file lacked some of the important channels and that it contained many empty channels.

I assumed that there is just one pycontrol txt file in a folder. If not, this will create some errors.

Test run on a single session looked good.

**Review by juliencarponcy — APPROVED** [review]
Did not tested but assume right. should we merge master yet? Or should we do the PR on dev?

**Comment by teristam** [comment]
I would suggest merging it to the `dev` branch, as the master branch doesn't support task-based folder structure, so there is no reliable way of testing it on the current folder structure. Please resolve the conflict before merging.

**Comment by teristam** [comment]
I will help resolve the conflict with dev. Please wait

**Review by teristam — APPROVED** [review]
Tested it and everything seems to look good. Also resolved the conflict with the dev branch. Merging.

**Comment by kouichi-c-nakamura** [comment]
Thanks! What now? Do you run it for everything at some point?

**Comment by teristam** [comment]
> Thanks! What now? Do you run it for everything at some point?

it is now merged to the `dev` branch. I tried running the pipeline on all our dataset but encountered errors due to the "ghost file" problem. I don't think that's related to this export feature. Guess we will have to find some other ways to solve the ghost file problem on the network drive..

**Comment by kouichi-c-nakamura** [comment]
Encountered an error when exporting 'reaching_go_spout_bar_all_reward_dec22', 'JC317L-2022-12-16-173145'

pulse_times_A len:90 pulse_times_B len: 16

RsyncError: Two pulse times differ too much. Results may not be correct

line68 in 03 script

Maybe need `try` to handle the error

It's likely that pycontrol and pyphotometry files are paired wrongly.

**Comment by teristam** [comment]
> Encountered an error when exporting 'reaching_go_spout_bar_all_reward_dec22', 'JC317L-2022-12-16-173145'
> 
> pulse_times_A len:90 pulse_times_B len: 16
> 
> RsyncError: Two pulse times differ too much. Results may not be correct
> 
> line68 in 03 script
> 
> Maybe need `try` to handle the error
> 
> It's likely that pycontrol and pyphotometry files are paired wrongly.

It is deliberately done that way so that the pipeline will refuse to proceed for the wrong file, otherwise, everything will be messed up. We can add the `-k` option when running snakemake to skip that particular recording. I suggest we open another issue for this as this is not related to this PR directly.

**Pull request merged**